src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java

```diff
@@ -0,0 +1,305 @@
+
+        // The input counts with rows (targets) sorted so that they match the PoN's order.
+        final RealMatrix tangentNormalizationRawInputCounts = targetMapper.fromCaseToPoNCounts(targetFactorNormalizedCounts.counts());
+
+        // We prepare the counts for tangent normalization.
+        final RealMatrix tangentNormalizationInputCounts = composeTangentNormalizationInputMatrix(tangentNormalizationRawInputCounts);
+
+        if (ctx == null) {
+            return tangentNormalizeNonSpark(targetFactorNormalizedCounts, reducedPanelCounts, reducedPanelPInvCounts, targetMapper, tangentNormalizationInputCounts);
+        } else {
+            return tangentNormalizeSpark(targetFactorNormalizedCounts, reducedPanelCounts, reducedPanelPInvCounts, targetMapper, tangentNormalizationInputCounts, ctx);
+        }
+    }
+
+    /**
+     * Tangent normalize given the raw PoN data without using Spark.
+     */
+    private static PCATangentNormalizationResult tangentNormalizeNonSpark(final ReadCountCollection targetFactorNormalizedCounts,
+                                                                          final RealMatrix reducedPanelCounts,
+                                                                          final RealMatrix reducedPanelPInvCounts,
+                                                                          final Case2PoNTargetMapper targetMapper,
+                                                                          final RealMatrix tangentNormalizationInputCounts) {
+        // Calculate the beta-hats for the input read count columns (samples).
+        logger.info("Calculating beta hats...");
+        final RealMatrix tangentBetaHats = calculateBetaHats(reducedPanelPInvCounts, tangentNormalizationInputCounts, EPSILON);
+
+        // Actual tangent normalization step.
+        logger.info("Performing actual tangent normalization (" + tangentNormalizationInputCounts.getColumnDimension() + " columns)...");
+        final RealMatrix tangentNormalizedCounts = tangentNormalize(reducedPanelCounts, tangentNormalizationInputCounts, tangentBetaHats);
+
+        // Output the tangent normalized counts.
+        logger.info("Post-processing tangent normalization results...");
+        final ReadCountCollection tangentNormalized = targetMapper.fromPoNtoCaseCountCollection(tangentNormalizedCounts, targetFactorNormalizedCounts.columnNames());
+        final ReadCountCollection preTangentNormalized = targetMapper.fromPoNtoCaseCountCollection(tangentNormalizationInputCounts, targetFactorNormalizedCounts.columnNames());
+
+        return new PCATangentNormalizationResult(tangentNormalized, preTangentNormalized, tangentBetaHats, targetFactorNormalizedCounts);
+    }
+
+    /**
+     * Calculate the beta-hats that best fit case read counts given the panel of normals.
+     *
+     * @param normalsInverse the log-normalized or reduced-panel pseudoinverse from a panel of normals
```

Comment:
also not happy about PInverse vs Pseudoinverse (PInverse sounds like the inverse of matrix P) either, but i'll live